src/server.rs

```diff
@@ -273,38 +273,36 @@ fn send_index(
     Ok(())
 }
 
-const HTREE_SEND_PIPELINE_LEN: usize = 5;
-
 fn send_htree(
     repo: &mut repository::Repo,
     tr: &mut htree::TreeReader,
     w: &mut dyn std::io::Write,
 ) -> Result<(), anyhow::Error> {
     let mut storage_engine = repo.storage_engine()?;
-
     loop {
         match tr.current_height() {
             Some(0) => {
-                let mut chunk_address = address::Address::default();
-                let mut pipeline = std::collections::VecDeque::new();
-
-                for ent_slice in tr.pop_level().unwrap().chunks(8 + address::ADDRESS_SZ) {
-                    let address_slice = &ent_slice[8..];
-                    chunk_address.bytes.clone_from_slice(&address_slice);
-                    pipeline.push_back((
-                        chunk_address,
-                        storage_engine.get_chunk_async(&chunk_address),
-                    ));
-
-                    if pipeline.len() >= HTREE_SEND_PIPELINE_LEN {
-                        let (chunk_address, chunk_data) = pipeline.pop_front().unwrap();
-                        write_chunk(w, &chunk_address, &chunk_data.recv()??)?;
-                    }
-                }
+                let address_buf = tr.pop_level().unwrap();
+
+                let mut addresses =
+                    Vec::with_capacity(address_buf.len() / (8 + address::ADDRESS_SZ));
+
+                // XXX Can we avoid this potentially large allocation/copy?
+                // It serves very little purpose and could double ram usage of this function
+                // in a bad case.
+                addresses.extend(
+                    address_buf
+                        .chunks(8 + address::ADDRESS_SZ)
+                        .map(|x| address::Address::from_slice(&x[8..]).unwrap()),
+                );
```

Comment:
If `pipeline_get_chunks` accepted an iterator of `Address` you could evaluate this lazily I guess?

Reply:
Yeah, I think this would be a good improvement, though it makes the plugin protocol a bit more complicated.

Reply:
It's easy enough to produce an iterator using https://doc.rust-lang.org/std/iter/fn.once.html, but if you just have a single value. I guess we could also make the trait have two methods and provide a default implementation that calls the other with `std::iter::once`.

Reply:
@runfalk are you familiar with s3? I would be ok with sharing the s3 code with you though its not in a state I want to make totally public.

Reply:
I know the concept of how it works, but I have never used it. You can find my e-mail in my commit history, if you want a second pair of eyes. I can't give much help on S3 specifics though.

Reply:
I'm not sure I follow why having an iterator makes the plugin protocol complicated? I looked at the S3 implementation and that only interacts with bupstash over a socket.

Reply:
Because the idea is to reduce the total ram usage, we need to avoid converting the iterator to a full Vec on either side of the socket, this means the protocol itself must support iteration of some kind.

Reply:
I see, that makes sense. I was naively talking about how to save memory on this side of the call site. I didn't look much further than the `pipelined_get_chunks` boundary.